File: app/src/main/java/com/sample/app/fragment_with_dialog/dialogs/DialogWithLayout.java
package com.sample.app.fragment_with_dialog.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;

import com.sample.app.R;

public class DialogWithLayout extends DialogFragment implements View.OnClickListener {

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        // Get the layout inflater
        LayoutInflater inflater = getActivity().getLayoutInflater();
        builder.setCancelable(false);
        View view = inflater.inflate(R.layout.dialog_layout, null);

        Button okButton = (Button) view.findViewById(R.id.ok_button);
        Button cancelButton = (Button) view.findViewById(R.id.cancel_button);
        okButton.setOnClickListener(this);
        cancelButton.setOnClickListener(this);

        builder.setView(view);

        return builder.create();
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.cancel_button:
                dismiss();
                break;
            case R.id.ok_button:
                dismiss();
                break;
        }
    }
}

File: app/src/main/java/com/sample/app/fragment_with_dialog/dialogs/DialogWithLayout.kt
package com.sample.app.fragment_with_dialog.dialogs

import android.app.Dialog
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v7.app.AlertDialog
import android.view.View
import android.widget.Button
import com.sample.app.R

class DialogWithLayout : DialogFragment(), View.OnClickListener {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(activity)

        // Get the layout inflater
        val inflater = activity.layoutInflater
        builder.setCancelable(false)
        val view = inflater.inflate(R.layout.dialog_layout, null)

        val okButton = view.findViewById(R.id.ok_button) as Button
        val cancelButton = view.findViewById(R.id.cancel_button) as Button
        okButton.setOnClickListener(this)
        cancelButton.setOnClickListener(this)

        builder.setView(view)

        return builder.create()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.cancel_button -> dismiss()
            R.id.ok_button -> dismiss()
        }
    }
}
